compare_platforms/include/review.h

```diff
@@ -0,0 +1,151 @@
+        rv->publicreview = publicreview;
+        // If we got this far, it was writable.
+        return true;
+    #else
+        if(can_write(rv)) {
+            rv->publicreview = publicreview;
+            return true;
+        }
+        else {
+            printf("You don't have permission to change this public review.\n");
+            return false;
+        }
+    #endif
+}
+
+/**
+ * A method for attempting to change the public review to something more flattering.
+ * param: rv, the struct representing the public review in question.
+ * param: reqsz, a size_t representing the amount of memory to allocate for the new review.
+ * param: b_weak, a boolean that is true if we are weakly enforcing (i.e. using can_write) the 
+ * read-only permissions, false otherwise.
+ * return: false if review is read-only, true if review is writable.
+**/
+bool try_improve_publicreview(struct review *rv, size_t reqsz, bool b_weak) {
+    bool b_writable = false;
+    
+    char *newpublicreview = malloc(reqsz);
+    strcpy(newpublicreview, "5 ***** strong accept. This author deserves two Nobels and an ice cream.\n");
+
+    printf("\nAttempting to change the public review.\n");
+    #ifdef __CHERI_PURE_CAPABILITY__
+        if((can_write(rv) == false) && (b_weak == true)) {
+            printf("You do not have permission to change the public review.\n");
+            return false;
+        } else {
+            // if the review is set to read-only, this will crash.
+            b_writable = change_publicreview(rv, newpublicreview);
+            return b_writable;
+        }
+    #else
+        b_writable = change_publicreview(rv, newpublicreview);
+        return b_writable;
+    #endif
+}
+
+/**
+ * Print all the member values for a given review.
+ * param: rv, the struct representing the review in question
+**/
+void print_details(struct review *rv)
+{
+    printf("\nReviewer Username: %s \n", rv->username);
+    printf("Reviewer's Real Name: %s \n", rv->realname);
+    printf("Private Review: %s\n", rv->privatereview);
+    printf("Public Review: %s\n", rv->publicreview);
+    printf("Review's Read-Write Permissions: %s\n", rv->permissions);
+    fflush(stdout);
+}
+
+/**
+ * Get the actual length of a string being assigned to a member.
+ * param: userdata, a char * to be assigned to a member of the review struct
+ * param: explen, an int representing the allocated length
+ * return: actsize, a size_t representing the actual size of the string
+ **/
+size_t actual_input_length(char *userdata, int explen) 
+{
+    size_t actsz = (size_t)strlen(userdata);
+    assert(actsz > 0);
+    assert(actsz <= explen);
+    
+    return actsz;
+}
+
+
+
+#ifdef __CHERI_PURE_CAPABILITY__
+/**
+ * Set all the struct members to read-only.
+ * param: rv, the struct representing the review to display
+ * return: a review with permissions restricted to read-only
+ **/
+struct review *set_read_only(struct review *rv)
```

Comment:
Hmm, this function being only defined in a CHERI environment might lead to issues during compilation. I would suggest we move the `ifdef` inside the function, and if we are not in a CHERI environment, we `assert(false)`.

Reply:
Done.

Reply:
> Hmm, this function being only defined in a CHERI environment might lead to issues during compilation. I would suggest we move the `ifdef` inside the function, and if we are not in a CHERI environment, we `assert(false)`.

Fixed in 84d32d463865d0ff38a349e845ffbed0ea7cd48a